programs/mango-v4/src/instructions/openbook_v2_place_order.rs

```diff
@@ -0,0 +1,565 @@
+        // Example: rebates go from 100 -> 10. That means we credit 90 in fees.
+        received_fees = before_oo
+            .native_rebates()
+            .saturating_sub(after_oo.native_rebates());
+        quote_bank.collected_fees_native += I80F48::from(received_fees);
+
+        // Credit the buyback_fees at the current value of the quote token.
+        if let Some(quote_oracle_ai) = quote_oracle {
+            let clock = Clock::get()?;
+            let now_ts = clock.unix_timestamp.try_into().unwrap();
+
+            let quote_oracle_ref = &AccountInfoRef::borrow(quote_oracle_ai)?;
+            let quote_oracle_price = quote_bank.oracle_price(
+                &OracleAccountInfos::from_reader(quote_oracle_ref),
+                Some(clock.slot),
+            )?;
+            let quote_asset_price = quote_oracle_price.min(quote_bank.stable_price());
+            account
+                .fixed
+                .expire_buyback_fees(now_ts, group.buyback_fees_expiry_interval);
+            let fees_in_usd = I80F48::from(received_fees) * quote_asset_price;
+            account
+                .fixed
+                .accrue_buyback_fees(fees_in_usd.clamp_to_u64());
+        }
+    }
+
+    // Don't count the referrer rebate fees as part of the vault change that should be
+    // credited to the user.
+    let after_quote_vault_adjusted = after_quote_vault - received_fees;
+
+    // Settle cannot decrease vault balances
+    require_gte!(after_base_vault, before_base_vault);
+    require_gte!(after_quote_vault_adjusted, before_quote_vault);
+
+    // Credit the difference in vault balances to the user's account
+    let base_difference = apply_vault_difference(
+        account_pk,
+        account,
+        openbook_market.market_index,
+        base_bank,
+        after_base_vault,
+        before_base_vault,
+    )?;
+    let quote_difference = apply_vault_difference(
+        account_pk,
+        account,
+        openbook_market.market_index,
+        quote_bank,
+        after_quote_vault_adjusted,
+        before_quote_vault,
+    )?;
+
+    // Tokens were moved from open orders into banks again: also update the tracking
+    // for potential_serum_tokens on the banks.
+    {
+        let openbook_orders = account.openbook_v2_orders_mut(openbook_market.market_index)?;
+
+        update_bank_potential_tokens_v2(openbook_orders, base_bank, quote_bank, after_oo);
+    }
+
+    if let Some(health_cache) = health_cache {
+        base_difference.adjust_health_cache_token_balance(health_cache, &base_bank)?;
+        quote_difference.adjust_health_cache_token_balance(health_cache, &quote_bank)?;
+
+        let serum_account = account.openbook_v2_orders(openbook_market.market_index)?;
+        OODifference::new(&before_oo, &after_oo).recompute_health_cache_openbook_v2_state(
+            health_cache,
+            serum_account,
+            open_orders,
+        )?;
+    }
+
+    Ok(())
+}
+
+fn update_bank_potential_tokens_v2(
```

Comment:
since this is a private function, the _v2 suffix confused me. probably you did this for consistency with `apply_settle_changes_v2` which is public? Maybe we could always refer to `openbook_v2_place_order::apply_settle_changes` explicitly and not `use ..:::*;` in lib.rs? hmm